pybamm/models/submodels/equivalent_circuit_elements/diffusion_element.py

```diff
@@ -0,0 +1,109 @@
+    equivalent circuits.
+
+    Parameters
+    ----------
+    param : parameter class
+        The parameters to use for this submodel
+    options : dict, optional
+        A dictionary of options to be passed to the model.
+    """
+
+    def __init__(self, param, options=None):
+        super().__init__(param)
+        self.model_options = options
+
+    def get_coupled_variables(self, variables):
+        z = pybamm.PrimaryBroadcast(variables["SoC"], "ECMD particle")
+        x = pybamm.SpatialVariable(
+            "x ECMD", domain=["ECMD particle"], coord_sys="Cartesian"
+        )
+        z_surf = pybamm.surf(z)
+        eta_diffusion = pybamm.Scalar(0)
+
+        variables.update(
+            {
+                "Distributed SoC": z,
+                "x ECMD": x,
+                "Diffusion overpotential [V]": eta_diffusion,
+                "Surface SoC": z_surf,
+            }
+        )
+
+        return variables
+
+
+class DiffusionElement(pybamm.BaseSubModel):
+    """
+    With Diffusion element for
+    equivalent circuits.
+
+    Parameters
+    ----------
+    param : parameter class
+        The parameters to use for this submodel
+    options : dict, optional
+        A dictionary of options to be passed to the model.
+    """
+
+    def __init__(self, param, options=None):
+        super().__init__(param)
+        self.model_options = options
+
+    def get_fundamental_variables(self):
+        z = pybamm.Variable("Distributed SoC", domain="ECMD particle")
+        x = pybamm.SpatialVariable(
+            "x ECMD", domain=["ECMD particle"], coord_sys="Cartesian"
+        )
+        variables = {
+            "Distributed SoC": z,
+            "x ECMD": x,
+        }
+        return variables
+
+    def get_coupled_variables(self, variables):
+        z = variables["Distributed SoC"]
+        soc = variables["SoC"]
+        z_surf = pybamm.surf(z)
+        eta_diffusion = -(self.param.ocv(z_surf) - self.param.ocv(soc))
+
+        variables.update(
+            {
+                "Diffusion overpotential [V]": eta_diffusion,
+                "Surface SoC": z_surf,
+            }
+        )
+
+        return variables
+
+    def set_rhs(self, variables):
+        cell_capacity = self.param.cell_capacity
+        current = variables["Current [A]"]
+        z = variables["Distributed SoC"]
+
+        # governing equations
+        dzdt = pybamm.div(pybamm.grad(z)) / self.param.tau_D
+        self.rhs = {z: dzdt}
+
+        # boundary conditions
+        lbc = pybamm.Scalar(0)
+        rbc = -self.param.tau_D * current / (cell_capacity * 3600)
+        self.boundary_conditions = {
+            z: {"left": (lbc, "Neumann"), "right": (rbc, "Neumann")}
+        }
+
+    def set_initial_conditions(self, variables):
+        z = variables["Distributed SoC"]
+        self.initial_conditions = {z: self.param.initial_soc}
+
+    def set_events(self, variables):
+        z_surf = variables["Surface SoC"]
+        self.events += [
```

Comment:
I just noticed that the same append behavior is in this version of the function. Same as below, at minimum the function name needs to change, but if there is not known issue then can it stay as a basic setter?

Reply:
I'm guessing @MehrdadBabazadeh followed how we do this in several other places, which I agree is the wrong way to do it. We're pretty haphazard throughout the submodels codebase about creating new dictionaries vs updating the pre-defined empty dictionaries. One way to enforce the better approach you suggest here would be to not initialize an empty dictionary at all

Reply:
Yes, the `set_events` function is pretty much everywhere. For this PR, I just suggested to Mehrdad to replace the `=` to a `+=` to avoid overwriting existing events. Going forward it would be probably better to have events as a dictionary, which would also make it easier to access specific events if the users want to change anything. But I would suggest doing this a separate PR.